components/tree/index.jsx

```diff
@@ -27,10 +27,97 @@ import { TREE } from '../../utilities/constants';
  * A tree is visualization of a structure hierarchy. A branch can be expanded or collapsed. This is a controlled component, since visual state is present in the `nodes` data.
  */
 class Tree extends React.Component {
+	constructor (props) {
+		super(props);
+		this.handleClick = this.handleClick.bind(this);
+		this.handleNodeBlur = this.handleNodeBlur.bind(this);
+		this.state = {
+			flattenedNodes: this.flattenTree({
+				nodes: this.props.nodes,
+				expanded: true,
+			}).slice(1),
+			selectedNodeIndexes: [],
+		};
+	}
+
 	componentWillMount () {
 		checkProps(TREE, this.props);
 	}
 
+	componentWillReceiveProps (nextProps) {
+		this.setState({
+			flattenedNodes: this.flattenTree({
+				nodes: nextProps.nodes,
+				expanded: true,
+			}).slice(1),
+		});
+	}
+
+	shouldComponentUpdate (nextProps, nextState) {
```

Comment:
I know this is kind of weird, but this is the simplest solution that I could think of. The problem that I was running into was that when clicking on a node, the blur is triggered and it would re-render the tree; as a result, the click event would never be triggered.